Flaky shouldRecoverAfterOutOfDiskSpaceAfterExporting: io.camunda.zeebe.it.health.DiskSpaceRecoveryIT$WithStandardContainerTes

**Summary**

> Try to answer the following as best as possible

- How often does the test fail?
- Does it block your work?
- Do we suspect that it is a real failure?

**Failures**

> Outline known failure cases, e.g. a failed assertion and its stacktrace obtained from Jenkins

<details><summary>Example assertion failure</summary>
<pre>
org.awaitility.core.ConditionTimeoutException: 
Condition with alias 'until the disk is not full anymore' didn't complete within 3 minutes because assertion condition with alias until the disk is not full anymore 
Expecting code not to raise a throwable but caught
  "io.camunda.zeebe.client.api.command.ClientStatusException: Cannot accept requests for partition 1. Broker is out of disk space
	at io.camunda.zeebe.client.impl.ZeebeClientFutureImpl.transformExecutionException(ZeebeClientFutureImpl.java:116)
	at io.camunda.zeebe.client.impl.ZeebeClientFutureImpl.join(ZeebeClientFutureImpl.java:54)
	at io.camunda.zeebe.it.health.DiskSpaceRecoveryIT.publishMessage(DiskSpaceRecoveryIT.java:84)
	at org.assertj.core.api.ThrowableAssert.catchThrowable(ThrowableAssert.java:63)
	at org.assertj.core.api.NotThrownAssert.isThrownBy(NotThrownAssert.java:43)
	at io.camunda.zeebe.it.health.DiskSpaceRecoveryIT$WithStandardContainerTest.lambda$shouldRecoverAfterOutOfDiskSpaceAfterExporting$1(DiskSpaceRecoveryIT.java:132)
	at org.awaitility.core.AssertionCondition.lambda$new$0(AssertionCondition.java:53)
	at org.awaitility.core.ConditionAwaiter$ConditionPoller.call(ConditionAwaiter.java:248)
	at org.awaitility.core.ConditionAwaiter$ConditionPoller.call(ConditionAwaiter.java:235)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:317)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)
	at java.base/java.lang.Thread.run(Thread.java:1583)
Caused by: java.util.concurrent.ExecutionException: io.grpc.StatusRuntimeException: RESOURCE_EXHAUSTED: Cannot accept requests for partition 1. Broker is out of disk space
	at java.base/java.util.concurrent.CompletableFuture.reportGet(CompletableFuture.java:396)
	at java.base/java.util.concurrent.CompletableFuture.get(CompletableFuture.java:2073)
	at io.camunda.zeebe.client.impl.ZeebeClientFutureImpl.join(ZeebeClientFutureImpl.java:52)
	... 11 more
Caused by: io.grpc.StatusRuntimeException: RESOURCE_EXHAUSTED: Cannot accept requests for partition 1. Broker is out of disk space
	at io.grpc.Status.asRuntimeException(Status.java:533)
	at io.grpc.stub.ClientCalls$StreamObserverToCallListenerAdapter.onClose(ClientCalls.java:481)
	at io.grpc.internal.ClientCallImpl.closeObserver(ClientCallImpl.java:574)
	at io.grpc.internal.ClientCallImpl.access$300(ClientCallImpl.java:72)
	at io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl$1StreamClosed.runInternal(ClientCallImpl.java:742)
	at io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl$1StreamClosed.runInContext(ClientCallImpl.java:723)
	at io.grpc.internal.ContextRunnable.run(ContextRunnable.java:37)
	at io.grpc.internal.SerializingExecutor.run(SerializingExecutor.java:133)
	... 3 more
".
	at org.awaitility.core.ConditionAwaiter.await(ConditionAwaiter.java:167)
	at org.awaitility.core.AssertionCondition.await(AssertionCondition.java:119)
	at org.awaitility.core.AssertionCondition.await(AssertionCondition.java:31)
	at org.awaitility.core.ConditionFactory.until(ConditionFactory.java:985)
	at org.awaitility.core.ConditionFactory.untilAsserted(ConditionFactory.java:769)
	at io.camunda.zeebe.it.health.DiskSpaceRecoveryIT$WithStandardContainerTest.shouldRecoverAfterOutOfDiskSpaceAfterExporting(DiskSpaceRecoveryIT.java:131)
	at java.base/java.lang.reflect.Method.invoke(Method.java:580)
	at java.base/java.util.concurrent.RecursiveAction.exec(RecursiveAction.java:194)
	at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:387)
	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1312)
	at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1843)
	at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1808)
	at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:188)
Caused by: java.lang.AssertionError: 
Expecting code not to raise a throwable but caught
  "io.camunda.zeebe.client.api.command.ClientStatusException: Cannot accept requests for partition 1. Broker is out of disk space
	at io.camunda.zeebe.client.impl.ZeebeClientFutureImpl.transformExecutionException(ZeebeClientFutureImpl.java:116)
	at io.camunda.zeebe.client.impl.ZeebeClientFutureImpl.join(ZeebeClientFutureImpl.java:54)
	at io.camunda.zeebe.it.health.DiskSpaceRecoveryIT.publishMessage(DiskSpaceRecoveryIT.java:84)
	at org.assertj.core.api.ThrowableAssert.catchThrowable(ThrowableAssert.java:63)
	at org.assertj.core.api.NotThrownAssert.isThrownBy(NotThrownAssert.java:43)
	at io.camunda.zeebe.it.health.DiskSpaceRecoveryIT$WithStandardContainerTest.lambda$shouldRecoverAfterOutOfDiskSpaceAfterExporting$1(DiskSpaceRecoveryIT.java:132)
	at org.awaitility.core.AssertionCondition.lambda$new$0(AssertionCondition.java:53)
	at org.awaitility.core.ConditionAwaiter$ConditionPoller.call(ConditionAwaiter.java:248)
	at org.awaitility.core.ConditionAwaiter$ConditionPoller.call(ConditionAwaiter.java:235)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:317)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)
	at java.base/java.lang.Thread.run(Thread.java:1583)
Caused by: java.util.concurrent.ExecutionException: io.grpc.StatusRuntimeException: RESOURCE_EXHAUSTED: Cannot accept requests for partition 1. Broker is out of disk space
	at java.base/java.util.concurrent.CompletableFuture.reportGet(CompletableFuture.java:396)
	at java.base/java.util.concurrent.CompletableFuture.get(CompletableFuture.java:2073)
	at io.camunda.zeebe.client.impl.ZeebeClientFutureImpl.join(ZeebeClientFutureImpl.java:52)
	... 11 more
Caused by: io.grpc.StatusRuntimeException: RESOURCE_EXHAUSTED: Cannot accept requests for partition 1. Broker is out of disk space
	at io.grpc.Status.asRuntimeException(Status.java:533)
	at io.grpc.stub.ClientCalls$StreamObserverToCallListenerAdapter.onClose(ClientCalls.java:481)
	at io.grpc.internal.ClientCallImpl.closeObserver(ClientCallImpl.java:574)
	at io.grpc.internal.ClientCallImpl.access$300(ClientCallImpl.java:72)
	at io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl$1StreamClosed.runInternal(ClientCallImpl.java:742)
	at io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl$1StreamClosed.runInContext(ClientCallImpl.java:723)
	at io.grpc.internal.ContextRunnable.run(ContextRunnable.java:37)
	at io.grpc.internal.SerializingExecutor.run(SerializingExecutor.java:133)
	... 3 more
"
	at io.camunda.zeebe.it.health.DiskSpaceRecoveryIT$WithStandardContainerTest.lambda$shouldRecoverAfterOutOfDiskSpaceAfterExporting$1(DiskSpaceRecoveryIT.java:132)
	at org.awaitility.core.AssertionCondition.lambda$new$0(AssertionCondition.java:53)
	at org.awaitility.core.ConditionAwaiter$ConditionPoller.call(ConditionAwaiter.java:248)
	at org.awaitility.core.ConditionAwaiter$ConditionPoller.call(ConditionAwaiter.java:235)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:317)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)
	at java.base/java.lang.Thread.run(Thread.java:1583)
</pre>
</details>

**Hypotheses**

> List any hypotheses if you have one; can be ommitted

**Logs**

https://github.com/camunda/zeebe/actions/runs/8024770580
[Test results for [IT] QA Integration Tests.zip](https://github.com/camunda/zeebe/files/14401732/Test.results.for.IT.QA.Integration.Tests.zip)

--- a/zeebe/qa/integration-tests/src/test/java/io/camunda/zeebe/it/health/DiskSpaceRecoveryIT.java
+++ b/zeebe/qa/integration-tests/src/test/java/io/camunda/zeebe/it/health/DiskSpaceRecoveryIT.java
@@ -112,10 +112,14 @@ final class DiskSpaceRecoveryIT {
           .atMost(Duration.ofMinutes(3))
           .pollInterval(1, TimeUnit.MICROSECONDS)
           .untilAsserted(
-              () ->
-                  assertThatThrownBy(DiskSpaceRecoveryIT.this::publishMessage)
-                      .hasRootCauseMessage(
-                          "RESOURCE_EXHAUSTED: Cannot accept requests for partition 1. Broker is out of disk space"));
+              () -> {
+                // Ensure at least one snapshot with a compactable processed position before going
+                // out of disk space
+                partitionsClient.takeSnapshot();
+                assertThatThrownBy(DiskSpaceRecoveryIT.this::publishMessage)
+                    .hasRootCauseMessage(
+                        "RESOURCE_EXHAUSTED: Cannot accept requests for partition 1. Broker is out of disk space");
+              });
 
       // when
       partitionsClient.resumeExporting();

--- a/zeebe/stream-platform/src/main/java/io/camunda/zeebe/stream/impl/StreamProcessor.java
+++ b/zeebe/stream-platform/src/main/java/io/camunda/zeebe/stream/impl/StreamProcessor.java
@@ -454,11 +454,8 @@ public class StreamProcessor extends Actor implements HealthMonitorable, LogReco
   public ActorFuture<Long> getLastProcessedPositionAsync() {
     return actor.call(
         () -> {
-          if (isInReplayOnlyMode()) {
+          if (isInReplayOnlyMode() || processingStateMachine == null) {
             return replayStateMachine.getLastSourceEventPosition();
-          } else if (processingStateMachine == null) {
-            // StreamProcessor is still replay mode
-            return StreamProcessor.UNSET_POSITION;
           } else {
             return processingStateMachine.getLastSuccessfulProcessedRecordPosition();
           }

--- a/zeebe/stream-platform/src/test/java/io/camunda/zeebe/stream/impl/StreamPlatform.java
+++ b/zeebe/stream-platform/src/test/java/io/camunda/zeebe/stream/impl/StreamPlatform.java
@@ -230,6 +230,12 @@ public final class StreamPlatform {
     return buildStreamProcessor(stream, false, StreamProcessorMode.REPLAY);
   }
 
+  public StreamProcessor startStreamProcessorPaused() {
+    final SynchronousLogStream stream = getLogStream();
+    return buildStreamProcessor(
+        stream, false, cfg -> cfg.streamProcessorMode(StreamProcessorMode.PROCESSING), true);
+  }
+
   public StreamProcessorLifecycleAware getMockProcessorLifecycleAware() {
     return mockProcessorLifecycleAware;
   }
@@ -255,6 +261,14 @@ public final class StreamPlatform {
       final SynchronousLogStream stream,
       final boolean awaitOpening,
       final Consumer<StreamProcessorBuilder> processorConfiguration) {
+    return buildStreamProcessor(stream, awaitOpening, processorConfiguration, false);
+  }
+
+  public StreamProcessor buildStreamProcessor(
+      final SynchronousLogStream stream,
+      final boolean awaitOpening,
+      final Consumer<StreamProcessorBuilder> processorConfiguration,
+      final boolean pauseOnStart) {
     final var storage = createRuntimeFolder(stream);
     final var snapshot = storage.getParent().resolve(SNAPSHOT_FOLDER);
     scheduledCommandCache = new TestCommandCache();
@@ -282,7 +296,7 @@ public final class StreamPlatform {
     builder.addLifecycleListener(mockProcessorLifecycleAware);
 
     final StreamProcessor streamProcessor = builder.build();
-    final var openFuture = streamProcessor.openAsync(false);
+    final var openFuture = streamProcessor.openAsync(pauseOnStart);
 
     if (awaitOpening) { // and recovery
       verify(mockProcessorLifecycleAware, timeout(15 * 1000)).onRecovered(any());

--- a/zeebe/stream-platform/src/test/java/io/camunda/zeebe/stream/impl/StreamProcessorReplayTest.java
+++ b/zeebe/stream-platform/src/test/java/io/camunda/zeebe/stream/impl/StreamProcessorReplayTest.java
@@ -9,6 +9,7 @@ package io.camunda.zeebe.stream.impl;
 
 import static io.camunda.zeebe.protocol.record.intent.ProcessInstanceIntent.ACTIVATE_ELEMENT;
 import static io.camunda.zeebe.protocol.record.intent.ProcessInstanceIntent.ELEMENT_ACTIVATING;
+import static org.assertj.core.api.Assertions.assertThat;
 import static org.mockito.ArgumentMatchers.any;
 import static org.mockito.Mockito.doAnswer;
 import static org.mockito.Mockito.doThrow;
@@ -25,10 +26,10 @@ import io.camunda.zeebe.protocol.record.intent.ErrorIntent;
 import io.camunda.zeebe.stream.api.ProcessingResultBuilder;
 import io.camunda.zeebe.stream.api.RecordProcessor;
 import io.camunda.zeebe.stream.api.records.TypedRecord;
+import io.camunda.zeebe.stream.impl.StreamProcessor.Phase;
 import io.camunda.zeebe.stream.util.RecordToWrite;
 import io.camunda.zeebe.stream.util.Records;
 import io.camunda.zeebe.test.util.junit.RegressionTest;
-import org.assertj.core.api.Assertions;
 import org.awaitility.Awaitility;
 import org.junit.jupiter.api.Test;
 import org.junit.jupiter.api.extension.ExtendWith;
@@ -153,19 +154,14 @@ final class StreamProcessorReplayTest {
 
     Awaitility.await("position has to be set on processing start")
         .untilAsserted(
-            () ->
-                Assertions.assertThat(streamProcessor.getLastProcessedPositionAsync().join())
-                    .isEqualTo(1L));
+            () -> assertThat(streamProcessor.getLastProcessedPositionAsync().join()).isEqualTo(1L));
     Awaitility.await("position has to be set on processing start")
         .untilAsserted(
-            () ->
-                Assertions.assertThat(streamProcessor.getLastWrittenPositionAsync().join())
-                    .isEqualTo(2L));
+            () -> assertThat(streamProcessor.getLastWrittenPositionAsync().join()).isEqualTo(2L));
 
     // state has to be updated
-    Assertions.assertThat(streamPlatform.getLastSuccessfulProcessedRecordPosition()).isEqualTo(1);
-    Assertions.assertThat(Protocol.decodeKeyInPartition(streamPlatform.getCurrentKey()))
-        .isEqualTo(19L);
+    assertThat(streamPlatform.getLastSuccessfulProcessedRecordPosition()).isEqualTo(1);
+    assertThat(Protocol.decodeKeyInPartition(streamPlatform.getCurrentKey())).isEqualTo(19L);
   }
 
   @Test
@@ -199,19 +195,14 @@ final class StreamProcessorReplayTest {
 
     Awaitility.await("position has to be set on processing start")
         .untilAsserted(
-            () ->
-                Assertions.assertThat(streamProcessor.getLastProcessedPositionAsync().join())
-                    .isEqualTo(1L));
+            () -> assertThat(streamProcessor.getLastProcessedPositionAsync().join()).isEqualTo(1L));
     Awaitility.await("position has to be set on processing start")
         .untilAsserted(
-            () ->
-                Assertions.assertThat(streamProcessor.getLastWrittenPositionAsync().join())
-                    .isEqualTo(2L));
+            () -> assertThat(streamProcessor.getLastWrittenPositionAsync().join()).isEqualTo(2L));
 
     // state has to be updated
-    Assertions.assertThat(streamPlatform.getLastSuccessfulProcessedRecordPosition()).isEqualTo(1);
-    Assertions.assertThat(Protocol.decodeKeyInPartition(streamPlatform.getCurrentKey()))
-        .isEqualTo(19L);
+    assertThat(streamPlatform.getLastSuccessfulProcessedRecordPosition()).isEqualTo(1);
+    assertThat(Protocol.decodeKeyInPartition(streamPlatform.getCurrentKey())).isEqualTo(19L);
   }
 
   @Test
@@ -250,24 +241,19 @@ final class StreamProcessorReplayTest {
     inOrder.verify(recordProcessor, TIMEOUT).replay(recordCaptor.capture());
     inOrder.verifyNoMoreInteractions();
 
-    Assertions.assertThat(recordCaptor.getValue().getKey()).isEqualTo(eventKeyAfterSnapshot);
+    assertThat(recordCaptor.getValue().getKey()).isEqualTo(eventKeyAfterSnapshot);
 
     Awaitility.await("position has to be set on processing start")
         .untilAsserted(
-            () ->
-                Assertions.assertThat(streamProcessor.getLastProcessedPositionAsync().join())
-                    .isEqualTo(3L));
+            () -> assertThat(streamProcessor.getLastProcessedPositionAsync().join()).isEqualTo(3L));
     Awaitility.await("position has to be set on processing start")
         .untilAsserted(
-            () ->
-                Assertions.assertThat(streamProcessor.getLastWrittenPositionAsync().join())
-                    .isEqualTo(4L));
-    Assertions.assertThat(Protocol.decodeKeyInPartition(streamPlatform.getCurrentKey()))
-        .isEqualTo(21L);
+            () -> assertThat(streamProcessor.getLastWrittenPositionAsync().join()).isEqualTo(4L));
+    assertThat(Protocol.decodeKeyInPartition(streamPlatform.getCurrentKey())).isEqualTo(21L);
   }
 
   @RegressionTest("https://github.com/camunda/zeebe/issues/13101")
-  public void shouldNotReplayErrorEventAppliedInSnapshot() throws Exception {
+  void shouldNotReplayErrorEventAppliedInSnapshot() throws Exception {
     // given
     final var processorWhichFails = setupProcessorWhichFailsOnProcessing();
     setupOnErrorReactionForProcessor(processorWhichFails);
@@ -296,6 +282,44 @@ final class StreamProcessorReplayTest {
     verifyProcessingErrorLifecycle(processorWhichFails);
     // we shouldn't replay any events - due to snapshot
     verify(processorWhichFails, never()).replay(any());
+  }
+
+  @Test
+  void shouldReturnLastProcessedPositionInSnapshotWhenPausedDuringReplay() throws Exception {
+    // given
+    final var positionInSnapshot =
+        streamPlatform.writeBatch(
+            RecordToWrite.command().processInstance(ACTIVATE_ELEMENT, Records.processInstance(1)));
+    streamPlatform.writeBatch(
+        RecordToWrite.event()
+            .processInstance(ELEMENT_ACTIVATING, Records.processInstance(1))
+            .key(1)
+            .causedBy((int) positionInSnapshot));
+    // starting the stream processor awaits until replay is completed
+    streamPlatform.startStreamProcessor();
+    streamPlatform.snapshot();
+
+    // write more events to make sure that the processor is paused in replay phase after recovery
+    streamPlatform.writeBatch(
+        RecordToWrite.command().processInstance(ACTIVATE_ELEMENT, Records.processInstance(1)),
+        RecordToWrite.event()
+            .processInstance(ELEMENT_ACTIVATING, Records.processInstance(1))
+            .key(2)
+            .causedBy(0));
+
+    streamPlatform.closeStreamProcessor();
+    streamPlatform.resetMockInvocations();
+
+    // when
+    final var streamProcessor = streamPlatform.startStreamProcessorPaused();
+
+    // then
+    assertThat(streamProcessor.getCurrentPhase().join()).isEqualTo(Phase.PAUSED);
+    Awaitility.await("position has to be set when replay is paused")
+        .untilAsserted(
+            () ->
+                assertThat(streamProcessor.getLastProcessedPositionAsync().join())
+                    .isEqualTo(positionInSnapshot));
   }
 
   private static void verifyProcessingErrorLifecycle(final RecordProcessor processorWhichFails) {
@@ -355,7 +379,6 @@ final class StreamProcessorReplayTest {
     streamPlatform.startStreamProcessor();
 
     // then
-    Assertions.assertThat(Protocol.decodeKeyInPartition(streamPlatform.getCurrentKey()))
-        .isEqualTo(19L);
+    assertThat(Protocol.decodeKeyInPartition(streamPlatform.getCurrentKey())).isEqualTo(19L);
   }
 }

